File: hw1/hmm.py
import math
import random
import numpy as np
import sys

try:
   from sys import maxint
except:
   maxint = 9223372036854775807

class HMM:
   def __init__(self, N, M, minIters, epsilon):
      self.N = N
      self.M = M
      self.minIters = minIters
      self.epsilon = epsilon
      self.oldLogProb = -maxint
      self.randomInit()

   def setupTable(self, T):
      self.c = np.zeros((T,), dtype=np.float32)
      self.alpha = np.zeros((T, self.N), dtype=np.float32)
      self.beta = np.zeros((T, self.N), dtype=np.float32)
      self.gamma = np.zeros((T, self.N), dtype=np.float32)
      self.diGamma = np.zeros((T, self.N, self.N), dtype=np.float32)

   def getRandVal(self, k): #return a random value around 1/k
      v = 1.0/k
      delta = 0.1 * v
      return v + (-delta + random.random() * 2 * delta);

   def getNorm(self, pList):
      s = sum(pList)
      norm = [float(p) / s for p in pList]
      #if sum(norm) != 1: norm[-1] = 1 - sum(norm[:-1])
      #if sum(norm) != 1.0: raise Exception("invalid normalized, norm %s, sum = %s" % (norm, sum(norm)))
      return norm

   def randomInit(self):
      #init A
      self.A = []
      for i in range(self.N):
         self.A.append([])
         for j in range(self.N):
            self.A[i].append(self.getRandVal(self.N))
         self.A[i] = self.getNorm(self.A[i])

      #init B
      self.B = []
      for i in range(self.N):
         self.B.append([])
         for j in range(self.M):
            self.B[i].append(self.getRandVal(self.M))
         self.B[i] = self.getNorm(self.B[i])

      #init PI
      self.PI = []
      for i in range(self.N): self.PI.append(self.getRandVal(self.N))
      self.PI = self.getNorm(self.PI)

      print("randomInit A = %s" % self.A)
      print("randomInit B = %s" % self.B)
      print("randomInit PI = %s" % self.PI)

   def forwardPass(self, obserSeq):
      #compute a0(i)
      self.c[0] = 0
      for i in range(self.N): 
          _a = self.PI[i] * self.B[i][obserSeq[0]]
          self.alpha[0][i] = _a
          self.c[0] = self.c[0] + _a 

      # scale the a0(i)
      self.c[0] = 1 / self.c[0] 
      for i in range(self.N):
         self.alpha[0][i] = self.alpha[0][i] * self.c[0]
      
      # compute at(i)
      for t in range(1, len(obserSeq)):
         self.c[t] = 0
         for i in range(self.N):
            self.alpha[t][i] = 0
            for j in range(self.N):
               self.alpha[t][i] = self.alpha[t][i] + (self.alpha[t-1][j]*self.A[j][i])
            self.alpha[t][i] = self.alpha[t][i] * self.B[i][obserSeq[t]]
            self.c[t] = self.c[t] + self.alpha[t][i]

         #scale at(i)
         self.c[t] = 1 / self.c[t]
         for i in range(self.N):
             self.alpha[t][i] = self.c[t] * self.alpha[t][i]


   def backwardPass(self, obserSeq):
      T = len(obserSeq)

      #let beta_t-1(i) = 1 scaled by cT-1
      for i in range(self.N):
         self.beta[T-1][i] = self.c[T-1]

      #beta-pass
      for t in reversed(range(T-1)):
         for i in range(self.N):
            self.beta[t][i] = 0
            for j in range(self.N):
               self.beta[t][i] = self.beta[t][i] + (self.A[i][j]*self.B[j][obserSeq[t+1]]*self.beta[t+1][j])

            # scale beta_ti with same scale factor as a_ti
            self.beta[t][i] = self.beta[t][i] * self.c[t] 

   def calcGammaDigamma(self, obserSeq):
      T = len(obserSeq)
      for t in range(T - 1):
         #denom = 0
         #for i in range(self.N):
         #   #self.gamma[t][i] = 0
         #   for j in range(self.N):
         #      #self.diGamma[t][i][j] = 0
         #      denom = denom + (self.alpha[t][i]*self.A[i][j]*self.B[j][obserSeq[t+1]]*self.beta[t+1][j])

         for i in range(self.N):
            self.gamma[t][i] = 0
            for j in range(self.N):
               #No need to normalize since using scaled alpha and beta
               self.diGamma[t][i][j] = (self.alpha[t][i]*self.A[i][j]*self.B[j][obserSeq[t+1]]*self.beta[t+1][j]) #/ denom
               self.gamma[t][i] = self.gamma[t][i] + self.diGamma[t][i][j]

      #special case for gamma_t-1(i)
      #denom = 0
      #for i in range(self.N): denom = denom + self.alpha[T-1][i]
      #No need to normalize since using scaled alpha and beta
      for i in range(self.N): self.gamma[T-1][i] = self.alpha[T-1][i] #/ denom


   def reEstimateModel(self, obserSeq):
      T = len(obserSeq)

      #re-estimate PI
      for i in range(self.N): self.PI[i] = self.gamma[0][i]
      #if sum(self.PI) != 1: self.PI[-1] = 1 - sum(self.PI[:-1])
      #if sum(self.PI) != 1: raise Exception("re-estimate PI invalid %s" % self.PI)

      #re-estimate A
      for i in range(self.N):
         for j in range(self.N):
            numer = 0
            denom = 0
            for t in range(T-1): 
               numer += self.diGamma[t][i][j]
               denom += self.gamma[t][i]
            if numer == 0: self.A[i][j] = 0 
            else: self.A[i][j] = numer / denom
         #if sum(self.A[i]) != 1: self.A[i][-1] = 1 - sum(self.A[i][:-1])
         #if sum(self.A[i]) != 1: raise Exception("re-estimate A invalid, %d %s", i, self.A[i])

      #re-estimate B
      for i in range(self.N):
         for j in range(self.M):
            numer = 0
            denom = 0
            for t in range(T):
               if obserSeq[t] == j: numer += self.gamma[t][i]
               denom += self.gamma[t][i]
            if numer == 0: self.B[i][j] = 0
            else: self.B[i][j] = numer / denom
         #if sum(self.B[i]) != 1: self.B[i][-1] = 1 - sum(self.B[i][:-1])
         #if sum(self.B[i]) != 1: raise Exception("re-estimate B invalid, %d, %s", i, self.B[i])


   def getScore(self, obserSeq):
      T = len(obserSeq)
      logProb = 0
      for t in range(T): 
         logProb += math.log(self.c[t]) 
      return -logProb


   def fit(self, obserSeq):
      if len(obserSeq) == 0: raise Exception("Cannot train HMM with empty seq")
      iters = 0
      logProb = 1 
      diff = maxint
      self.setupTable(len(obserSeq))
      while iters < self.minIters or diff > self.epsilon:
         self.forwardPass(obserSeq)
         self.backwardPass(obserSeq)
         self.calcGammaDigamma(obserSeq)
         self.reEstimateModel(obserSeq)
         logProb = self.getScore(obserSeq)
         #if iters % 10 == 0: print("%d: score = %s" % (iters, logProb))
         print("%d: score = %s" % (iters, logProb))
         diff = abs(logProb - self.oldLogProb)
         self.oldLogProb = logProb
         iters += 1
      return self.A, self.B, self.PI


def main():
   #hyperparameters:
   observationSet = 'abcdefghijklmnopqrstuvwxyz '
   N = 2
   if len(sys.argv) > 1: N = int(sys.argv[1])
   if N < 1: raise Exception("Invalid N %d" % N)

   M = len(observationSet)
   T = 50000
   minIterations = 20
   epsilon = 0.01
   
   #first, read the entire txt
   print("reading the text file")
   f = open('brown.txt', 'r')
   txt = f.read()
   f.close()

   #then, prepared the observation sequence
   print("preparing the observation sequence")
   obserSeq = []
   for ch in txt:
      obser = observationSet.find(ch.lower())
      if obser >= 0: obserSeq.append(obser)
      if len(obserSeq) >= T: break

   #start training an HMM
   print("start HMM training, unique observation str = %s" % observationSet)
   print("using N = %d, M = %d, T = %d, min iterations = %d, epsilon = %s" % (N, M, T, minIterations, epsilon))

   #create an HMM
   hmm = HMM(N, M, minIterations, epsilon) 

   #fit the HMM with obserSeq
   A, B, PI = hmm.fit(obserSeq)

   #HMM training finished
   print("HMM training finished. Model:")
   print("====== A ======", A)
   #for i in range(N): print("state %d = %s" % (i, A[i]))
   print("====== B ======", B)
   #for i in range(N): print("state %d = %s" % (i, B[i]))
   print("PI = %s" % PI)

if __name__ == '__main__':
   main()

File: hw2/hmm.cpp
#include <stdio.h>
#include <ctime>
#include <cstdlib>
#include <cmath>
#include <string>
#include <fstream>
#include "hmm.h"

#define MAX_INT 9223372036854775807

static const std::string ObserSet = "abcdefghijklmnopqrstuvwxyz";

HMM::HMM(int N, int M, int minIters, float epsilon)
{
   mN = N;
   mM = M;
   mMinIters = minIters;
   mEps = epsilon;
   mOldLogProb = -MAX_INT;

   mA = NULL;
   mB = NULL;
   mPI = NULL;

   mC = NULL;
   mAlpha = NULL;
   mBeta = NULL;
   mGamma = NULL;
   mDiGamma = NULL;
}
HMM::~HMM()
{
   freeTables();
}

void HMM::freeTables()
{
   if (mA) {
      free(mA);
      mA = NULL;
   }
   if (mB) {
      free(mB);
      mB = NULL;
   }
   if (mPI) {
      free(mPI);
      mPI = NULL;
   }
   if (mC) {
      free(mC);
      mC = NULL;
   }
   if (mAlpha) {
      free(mAlpha);
      mAlpha = NULL;
   }
   if (mBeta) {
      free(mBeta);
      mBeta = NULL;
   }
   if (mGamma) {
      free(mGamma);
      mGamma = NULL;
   }
   if (mDiGamma) {
      free(mDiGamma);
      mDiGamma = NULL;
   }
}

void HMM::printModel()
{
   printf("A = \n");
   for (int i = 0; i < mN; i++) {
      std::string msg = std::to_string(i) + ": ";
      for (int j = 0; j < mN; j++) {
         msg += std::to_string(*getA(i, j)) + " ";
      }
      printf("%s\n", msg.c_str());
   }
   printf("B = \n");
   for (int j = 0; j < mM; j++) {
      std::string msg = std::string(1, ObserSet[j]) + ": ";
      for (int i = 0; i < mN; i++) {
         msg += std::to_string(*getB(i, j)) + " ";
      }
      printf("%s\n", msg.c_str());
   }
   printf("PI = \n");
   std::string msg = "";
   for (int i = 0; i < mN; i++) {
      msg += std::to_string(*getPI(i)) + " ";
   }
   printf("%s\n", msg.c_str());
}

void HMM::fit(int* obserArr, int T)
{
   int iters = 0;
   float logProb = -MAX_INT;
   float diff = MAX_INT;
   setupTable(T);
   randomInit();
   printModel();
   while (iters < mMinIters || diff > mEps) {
      logProb = getScore(obserArr, T);
      backwardPass(obserArr, T);
      calcGammaDigamma(obserArr, T);
      reEstimateModel(obserArr, T);

      diff = std::abs(logProb - mOldLogProb); 
      mOldLogProb = logProb;

      //if (iters % 10 == 0) {
         printModel();
         printf("%d_score = %.3f\n========================================\n\n", iters, logProb);
      //}

      iters += 1;
   }

   printf("HMM trained\n");
   printModel();
   printf("%d_score = %.3f\n========================================\n\n", iters, logProb);
}

float HMM::getScore(int* obserArr, int T)
{
   forwardPass(obserArr, T);
   float logProb = 0;
   for (int t = 0; t < T; t++) logProb += std::log(*(getC(t))); 
   return -logProb;
}

void HMM::setupTable(int T)
{
   freeTables();

   mA = (float*)malloc(mN*mN*sizeof(float));
   if (!mA) {
      printf("No memory!\n");
      exit(-1);
   }
   mB = (float*)malloc(mN*mM*sizeof(float));
   if (!mB) {
      printf("No memory!\n");
      exit(-1);
   }
   mPI = (float*)malloc(mN*sizeof(float));
   if (!mPI) {
      printf("No memory!\n");
      exit(-1);
   }

   mC = (float*)malloc(T*sizeof(float));
   if (!mC) {
      printf("No memory!\n");
      exit(-1);
   }
   mAlpha = (float*)malloc(T*mN*sizeof(float));
   if (!mAlpha) {
      printf("No memory!\n");
      exit(-1);
   }
   mBeta = (float*)malloc(T*mN*sizeof(float));
   if (!mBeta) {
      printf("No memory!\n");
      exit(-1);
   }
   mGamma = (float*)malloc(T*mN*sizeof(float));
   if (!mGamma) {
      printf("No memory!\n");
      exit(-1);
   }
   mDiGamma = (float*)malloc(T*mN*mN*sizeof(float));
   if (!mDiGamma) {
      printf("No memory!\n");
      exit(-1);
   }
}
float HMM::getRandVal(int k)
{
   float r = float(rand())/RAND_MAX;
   float v = 1.0/k; 
   float delta = 0.1 * v;
   return v + -delta + (r * 2 * delta);
} 

void HMM::normalizeArr(float* arr, int T)
{
   float sum = 0;
   for (int idx = 0; idx < T; idx++) sum += arr[idx];
   if (sum == 0) {
      printf("error! can't normalize all 0\n");
      exit(-1);
   }
   for (int idx = 0; idx < T; idx++) arr[idx] /= sum;
}
void HMM::randomInit()
{
   srand(time(NULL));
   for (int idx = 0; idx < mN; idx++) {
      for (int jdx = 0; jdx < mN; jdx++) {
         *getA(idx, jdx) = getRandVal(mN);
      }
      normalizeArr(mA+(idx*mN), mN);
   }
   for (int idx = 0; idx < mN; idx++) {
      for (int jdx = 0; jdx < mM; jdx++) {
         *getB(idx, jdx) = getRandVal(mM);
      }
      normalizeArr(mB+(idx*mM), mM);
   }
   for (int idx = 0; idx < mN; idx++) {
      *getPI(idx) = getRandVal(mN);
   }
   normalizeArr(mPI, mN);
} 

void HMM::forwardPass(int* obserArr, int T)
{
   //compute a0[i]
   *getC(0) = 0;
   for (int idx = 0; idx < mN; idx++) {
      float _a = (*getPI(idx)) * (*getB(idx, obserArr[0])); 
      *getAlpha(0, idx) = _a;
      *getC(0) = (*getC(0)) + _a;
   }

   //scale the a0(i)
   *getC(0) = 1.0 / (*getC(0));
   for (int idx = 0; idx < mN; idx++) {
      *getAlpha(0, idx) = (*getAlpha(0, idx)) * (*getC(0));
   }

   //compute at(i)
   for (int t = 1; t < T; t++) {
      *getC(t) = 0;
      for (int idx = 0; idx < mN; idx++) {
         *getAlpha(t, idx) = 0;
         for (int jdx = 0; jdx < mN; jdx++) {
            *getAlpha(t, idx) = (*getAlpha(t, idx)) +  ((*getAlpha(t-1, jdx)) * (*getA(jdx, idx)));
         }
         *getAlpha(t, idx) = (*getAlpha(t, idx)) * (*getB(idx, obserArr[t]));
         *getC(t) = (*getC(t)) + (*getAlpha(t, idx));
      }

      //scale at(i)
      *getC(t) = 1.0 / (*getC(t));
      for (int idx = 0; idx < mN; idx++) {
         *getAlpha(t, idx) = (*getAlpha(t, idx)) * (*getC(t));
      }
   }
}

void HMM::backwardPass(int* obserArr, int T)
{
   //let beta_t-1(i) = 1 scaled by cT-1
   for (int idx = 0; idx < mN; idx++) {
      *getBeta(T-1, idx) = *getC(T-1);
   }

   //beta-pass
   for (int t = T-2; t > 0; t--) {
      for (int idx = 0; idx < mN; idx++) {
         *getBeta(t, idx) = 0;
         for (int jdx = 0; jdx <mN; jdx++) {
            *getBeta(t, idx) += (*getA(idx, jdx) * (*getB(jdx, obserArr[t+1])) * (*getBeta(t+1, jdx)));
         }

         //scale beta_ti with same scale factor as a_ti
         *getBeta(t, idx) = (*getBeta(t, idx)) * (*getC(t));
      }
   }
}

void HMM::calcGammaDigamma(int* obserArr, int T)
{
   for (int t = 0; t < T-1; t++) {
      for (int idx = 0; idx < mN; idx++) {
         *getGamma(t, idx) = 0;
         for (int jdx = 0; jdx < mN; jdx++) {
            //No need to normalize since using scaled alpha and beta 
            *getDiGamma(t, idx, jdx) = (*getAlpha(t, idx)) * (*getA(idx, jdx)) * (*getB(jdx, obserArr[t+1])) * (*getBeta(t+1, jdx));
            *getGamma(t, idx) = (*getGamma(t, idx)) + (*getDiGamma(t, idx, jdx));
         }
      }
   }

   //special case for gamma_t-1(i)
   //No need to normalize since using scaled alpha and beta 
   for (int idx = 0; idx < mN; idx++) *getGamma(T-1, idx) = *getAlpha(T-1, idx);
  
}

void HMM::reEstimateModel(int* obserArr, int T)
{
   //re-estimate PI
   for (int idx = 0; idx < mN; idx++) *getPI(idx) = *getGamma(0, idx);

   //re-estimate A
   for (int idx = 0; idx < mN; idx++) {
      for (int jdx = 0; jdx < mN; jdx++) {
         float numer = 0;
         float denom = 0;
         for (int t = 0; t < T-1; t++) {
            numer += *getDiGamma(t, idx, jdx);
            denom += *getGamma(t, idx);
         }
         if (numer == 0) *getA(idx, jdx) = 0;
         else *getA(idx, jdx) = numer / denom;
      }
   }

   //re-estimate B
   for (int idx = 0; idx < mN; idx++) {
      for (int jdx = 0; jdx < mM; jdx++) {
         float numer = 0;
         float denom = 0;
         for (int t = 0; t < T; t++) {
            if (obserArr[t] == jdx) numer += *getGamma(t, idx);
            denom += *getGamma(t, idx);
         }
         if (numer == 0) *getB(idx, jdx) = 0;
         else *getB(idx, jdx) = numer / denom;
      }
   }
}


float* HMM::getA(int i, int j)
{
   return mA + (i*mN) + j;
}
float* HMM::getB(int i, int j)
{
   return mB + (i*mM) + j;
}
float* HMM::getPI(int i)
{
   return mPI + i; 
}

float* HMM::getC(int t)
{
   return mC + t; 
}
float* HMM::getAlpha(int t, int i)
{
   return mAlpha + (t*mN) + i; 
}
float* HMM::getBeta(int t, int i)
{
   return mBeta + (t*mN) + i;
}
float* HMM::getGamma(int t, int i)
{
   return mGamma + (t*mN) + i;
}
float* HMM::getDiGamma(int t, int i, int j)
{
   return mDiGamma + (t*mN*mN) + (i*mN) + j;
}


int main(int argc, const char** argv) {
   if (argc != 5) {
      printf("Usage: %s <N> <txt> <T> <minIters>\n", argv[0]);
      return -1;
   }

   const int N = std::stoi(argv[1]);
   const int M = ObserSet.length();
   int T = std::stoi(argv[3]);
   const int minIters = std::stoi(argv[4]);
   const float epsilon = 0.00001;
   int* obsers = (int*)malloc(T*sizeof(int));
   if (!obsers) {
      printf("no memory\n");
      return -2;
   }

   //prepare observation sequence from txt file
   printf("using txt file %s\n", argv[2]);
   std::ifstream is(argv[2], std::ifstream::in);
   char c;
   int i = 0;
   while (is.get(c)) {
      int idx = ObserSet.find(c);
      if (idx == std::string::npos) continue;
      obsers[i] = idx;
      i++;
      if (i >= T) break;
   }
   is.close();
   if (i < T) {
      printf("file %s has %d less than %d chars\n", argv[2], i, T);
      T = i;
   }

   //start training
   HMM* hmm = new HMM(N, M, minIters, epsilon);
   if (!hmm) {
      printf("no memory\n");
      return -3;
   }

   printf("start training HMM for seq N = %d, M = %d, minIters = %d, eps = %.6f, T = %d\n", N, M, minIters, epsilon, T);
   hmm->fit(obsers, T);

   //clean up and exit
   delete hmm;
   free(obsers);
   return 0;
}



